ipykernel/iostream.py

```diff
@@ -105,87 +109,183 @@ def _check_mp_mode(self):
+    flush_interval = 0.2
+    topic=None
 
-            if data:
-                content = {u'name':self.name, u'text':data}
-                msg = self.session.send(self.pub_socket, u'stream', content=content,
-                                       parent=self.parent_header, ident=self.topic)
+    def __init__(self, session, pub_thread, name, pipe=None):
+        if pipe is not None:
+            warnings.warn("pipe argument to OutStream is deprecated and ignored",
+                DeprecationWarning)
+        self.encoding = 'UTF-8'
+        self.session = session
+        if not isinstance(pub_thread, IOPubThread):
+            # Backward-compat: given socket, not thread. Wrap in a thread.
+            warnings.warn("OutStream should be created with IOPubThread, not %r" % pub_thread,
+                DeprecationWarning, stacklevel=2)
+            pub_thread = IOPubThread(pub_thread)
+            pub_thread.start()
+        self.pub_thread = pub_thread
+        self.name = name
+        self.topic = b'stream.' + py3compat.cast_bytes(name)
+        self.parent_header = {}
+        self._master_pid = os.getpid()
+        self._flush_lock = threading.Lock()
+        self._flush_timeout = None
+        self._io_loop = pub_thread.io_loop
+        self._new_buffer()
 
-                if hasattr(self.pub_socket, 'flush'):
-                    # socket itself has flush (presumably ZMQStream)
-                    self.pub_socket.flush()
+    def _is_master_process(self):
+        return os.getpid() == self._master_pid
+
+    def _check_mp_mode(self):
+        """check for forks, and switch to zmq pipeline if necessary"""
+        if self._is_master_process():
+            return MASTER
         else:
-            with self._pipe_out_lock:
-                string = self._flush_buffer()
-                tracker = self._pipe_out.send_multipart([
-                    self._pipe_uuid,
-                    string.encode(self.encoding, 'replace'),
-                ], copy=False, track=True)
-                try:
-                    tracker.wait(1)
-                except:
-                    pass
+            return CHILD
 
+    def set_parent(self, parent):
+        self.parent_header = extract_header(parent)
+
+    def close(self):
+        self.pub_thread = None
+
+    @property
+    def closed(self):
+        return self.pub_thread is None
+
+    def _schedule_flush(self):
+        """schedule a flush in the IO thread
+        
+        call this on write, to indicate that flush should be called soon.
+        """
+        with self._flush_lock:
+            if self._flush_timeout is not None:
+                return
+            # non-None to avoid races while waiting for _schedule
+            self._flush_timeout = 'placeholder'
+        
+        # add_timeout has to be handed to the io thread with add_callback
+        def _schedule_in_thread():
+            # FIXME: call_later is a little nicer, if we can require tornado ≥ 4
```

Comment:
The *kernel* doesn't require tornado at all, so this is only necessary if IOLoop is coming from the bundled version in pyzmq. But given how ~everybody installs it, the odds of not having tornado are slim. I only noticed because we don't install tornado on travis for this repo.